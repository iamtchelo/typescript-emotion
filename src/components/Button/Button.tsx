import * as React from 'react';
import { Button } from './button.elements';

export interface ButtonProps {
  onClick?: (event: any) => void;
  actived?: boolean;
}

const ButtonSFC: React.SFC<ButtonProps> = ({
  onClick,
  actived,
  children
}) => (
  <Button
    onClick={onClick}
    actived={actived}
  >
    {children}
  </Button>
);

export default ButtonSFC;