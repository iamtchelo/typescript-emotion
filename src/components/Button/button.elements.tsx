import styled, { css } from 'react-emotion';
import { ButtonProps } from './Button';

export const Button = styled('button')`
  padding: 10px 40px;
  outline: none;
  cursor: pointer;
  font-size: 14px;
  border: 1px solid #7769A2;
  color: #414141;
  background: white;
  transition: background 0.1s linear, color 0.2s ease-out;

  &:hover {
    background: #DAD1F8;
  }

  ${({ actived }: ButtonProps) => actived && css`
    &, &:hover {
      background: #4E3C88;
      color: white;
    }
  `};
`;