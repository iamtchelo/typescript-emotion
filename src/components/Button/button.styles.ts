import styled, {css} from 'react-emotion';

const buttonBase = ({theme}: any) => css`
  padding: 10px 40px;
  font-size: 14px;
  outline: none;
  cursor: pointer;
  color: ${theme.button.color};
  background-color: ${theme.button.background};
  border: 1px solid ${theme.button.borderColor};
  transition: background-color 150ms linear, color 250ms ease-out;

  &:hover {
    background-color: ${theme.button.backgroundHover};
    color: ${theme.colors.white};
  }
`;

const buttonActived = ({theme, actived}: any) =>
  actived &&
  css`
    &,
    &:hover {
      background-color: ${theme.button.backgroundActived};
      color: ${theme.colors.white};
    }
  `;

const ButtonStyled = styled('button')`
  ${buttonBase};
  ${buttonActived};
`;

export {ButtonStyled};
