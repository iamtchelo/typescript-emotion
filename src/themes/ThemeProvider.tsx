import { ThemeProvider as Provider } from 'emotion-theming';
import * as React from 'react';
import { ComponentEnhancer, compose, setDisplayName ,withState } from 'recompose';
import primaryTheme from './primary';
import secondaryTheme from './secondary';

const themes = {
  primary: primaryTheme,
  secondary: secondaryTheme
};

export interface ThemeProviderProps {
  theme?: 'primary' | 'secondary';
  children?: React.SFC<any>;
};

export interface ThemeProviderDispatches {
  setTheme: (theme: 'primary' | 'secondary') => void;
};

export type EnhancedThemeProviderProps = ThemeProviderProps & ThemeProviderDispatches;

const ThemeProvider: React.SFC<EnhancedThemeProviderProps> = ({
  theme = 'primary',
  setTheme,
  children
}) => (
  <Provider theme={themes[theme]}>
    {children && children({ theme, setTheme })}
  </Provider>
);

const withTheme = compose(
  withState('theme', 'setTheme', ({ theme = 'primary' }: any) => theme)
);

const enhance = compose(
  setDisplayName('ThemeProvider'),
  withTheme
) as ComponentEnhancer<EnhancedThemeProviderProps, ThemeProviderProps>

export default enhance(ThemeProvider);
