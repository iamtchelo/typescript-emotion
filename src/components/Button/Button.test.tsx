import { render } from 'enzyme';
import { renderToJson } from 'enzyme-to-json';
import * as React from 'react';
import ThemeProvider from '../../themes';
import Button from './Button';

describe('<Button />', () => {
  test('should render with default style', () => {
    const tree = render(
      <ThemeProvider>
        {(props: any) => <Button {...props}>My button</Button>}
      </ThemeProvider>
    );
    expect(renderToJson(tree)).toMatchSnapshot();
  });

  test('should render with actived style', () => {
    const tree = render(
      <ThemeProvider>
        {(props: any) => <Button actived={true} {...props}>My button</Button>}
      </ThemeProvider>
    );
    expect(renderToJson(tree)).toMatchSnapshot();
  });
});