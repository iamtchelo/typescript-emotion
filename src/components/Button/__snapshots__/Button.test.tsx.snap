// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should render with actived style 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "css-1k3qov9",
    },
    "children": Array [
      Object {
        "data": "My button",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`<Button /> should render with default style 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "css-1k3qov9",
    },
    "children": Array [
      Object {
        "data": "My button",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "text",
      },
    ],
    "name": "button",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
